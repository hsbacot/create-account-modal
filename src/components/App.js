import React from "react";
import CreateAccountForm from "./CreateAccountForm";
import Modal from "./Modal";
import { validatePassword, validateEmail } from "../utils";

const initialState = {
  email: "",
  password: "",
  showCreateAccount: false,
  showEmailErrors: false,
  passwordValidations: validatePassword(""),
  emailValidations: {
    valid: false,
    errors: []
  }
};

class App extends React.Component {
  state = { ...initialState };

  showModal = () => {
    this.setState({ showCreateAccount: true });
  };

  hideModal = () => {
    this.setState({ showCreateAccount: false });
  };

  handlePassword = e => {
    this.setState({
      password: e.target.value,
      passwordValidations: validatePassword(e.target.value)
    });
  };

  handleEmail = e => {
    this.setState({
      email: e.target.value,
      emailValidations: validateEmail(e.target.value)
    });
  };

  handleSubmit = () => {
    this.setState({ ...initialState, userCreated: true });
  };

  displayEmailErrors = () => {
    this.setState({
      showEmailErrors: true
    });
  };

  render() {
    const {
      showCreateAccount,
      emailValidations,
      showEmailErrors,
      passwordValidations,
      password,
      email,
      userCreated
    } = this.state;

    // only show email after passwor
    // password at 8 characters
    const formErrors = [
      ...(showEmailErrors ? emailValidations.errors : []),
      ...passwordValidations.errors
    ];

    return (
      <div>
        {userCreated
          ? <p>Successfully created user</p>
          : <button className="btn btn-default" onClick={this.showModal}>
              Create Account
            </button>}
        {showCreateAccount &&
          <Modal>
            <h2>Create Account</h2>
            <CreateAccountForm
              email={email}
              password={password}
              closeModal={this.hideModal}
              formErrors={formErrors}
              handleEmail={this.handleEmail}
              handlePassword={this.handlePassword}
              handleSubmit={this.handleSubmit}
              displayEmailErrors={this.displayEmailErrors}
            />
            <button
              className="btn btn-default close-modal"
              onClick={this.hideModal}
            >
              Close
            </button>
          </Modal>}
      </div>
    );
  }
}

export default App;
