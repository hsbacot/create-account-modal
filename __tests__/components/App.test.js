import React from "react";
import { shallow } from "enzyme";
import App from "../../src/components/App";

const validEmailInput = {
  target: {
    value: "test@test.com"
  }
};

const validPasswordInput = {
  target: {
    value: "Pa$$word!"
  }
};

describe("<App />", () => {
  it("renders without crashing", () => {
    const component = shallow(<App />);
    expect(component).toMatchSnapshot();
  });

  it("should show and hide the modal", () => {
    let component = shallow(<App />).instance();
    component.showModal();
    expect(component).toMatchSnapshot();
    component.hideModal();
    expect(component).toMatchSnapshot();
  });

  it("should handle email input", () => {
    let component = shallow(<App />).instance();
    component.handleEmail(validEmailInput);
    expect(component).toMatchSnapshot();
  });

  it("should handle password input", () => {
    let component = shallow(<App />).instance();
    component.handlePassword(validPasswordInput);
    expect(component).toMatchSnapshot();
  });

  it("should know when to display email errors", () => {
    let component = shallow(<App />).instance();
    component.displayEmailErrors();
    expect(component).toMatchSnapshot();
  });

  it("should handle the form submit", () => {
    let component = shallow(<App />).instance();
    component.handleSubmit();
    expect(component).toMatchSnapshot();
  });
});
