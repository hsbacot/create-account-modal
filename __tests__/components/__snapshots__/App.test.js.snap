exports[`<App /> renders without crashing 1`] = `
<div>
  <button
    className="btn btn-default"
    onClick={[Function]}>
    Create Account
  </button>
</div>
`;

exports[`<App /> should handle email input 1`] = `
App {
  "_reactInternalInstance": ShallowComponentWrapper {
    "_calledComponentWillUnmount": false,
    "_compositeType": 0,
    "_context": Object {},
    "_currentElement": <App />,
    "_debugID": 5,
    "_hostContainerInfo": null,
    "_hostParent": null,
    "_instance": [Circular],
    "_mountOrder": 3,
    "_pendingCallbacks": null,
    "_pendingElement": null,
    "_pendingForceUpdate": false,
    "_pendingReplaceState": false,
    "_pendingStateQueue": null,
    "_renderedComponent": NoopInternalComponent {
      "_currentElement": <div>
        <button
                className="btn btn-default"
                onClick={[Function]}>
                Create Account
        </button>
</div>,
      "_debugID": 6,
      "_renderedOutput": <div>
        <button
                className="btn btn-default"
                onClick={[Function]}>
                Create Account
        </button>
</div>,
    },
    "_renderedNodeType": 0,
    "_rootNodeID": 0,
    "_topLevelWrapper": null,
    "_updateBatchNumber": null,
    "_warnedAboutRefsInRender": false,
  },
  "context": Object {},
  "displayEmailErrors": [Function],
  "handleEmail": [Function],
  "handlePassword": [Function],
  "handleSubmit": [Function],
  "hideModal": [Function],
  "props": Object {},
  "refs": Object {},
  "showModal": [Function],
  "state": Object {
    "email": "test@test.com",
    "emailValidations": Object {
      "errors": Array [],
      "valid": true,
    },
    "password": "",
    "passwordValidations": Object {
      "errors": Array [
        "Password must contain one of the following symbols: !#$%&\'*+-/=?^_\`{|}~",
        "Password must be at least 8 characters",
      ],
      "valid": false,
    },
    "showCreateAccount": false,
    "showEmailErrors": false,
  },
  "updater": Object {
    "enqueueCallback": [Function],
    "enqueueCallbackInternal": [Function],
    "enqueueElementInternal": [Function],
    "enqueueForceUpdate": [Function],
    "enqueueReplaceState": [Function],
    "enqueueSetState": [Function],
    "isMounted": [Function],
    "validateCallback": [Function],
  },
}
`;

exports[`<App /> should handle password input 1`] = `
App {
  "_reactInternalInstance": ShallowComponentWrapper {
    "_calledComponentWillUnmount": false,
    "_compositeType": 0,
    "_context": Object {},
    "_currentElement": <App />,
    "_debugID": 7,
    "_hostContainerInfo": null,
    "_hostParent": null,
    "_instance": [Circular],
    "_mountOrder": 4,
    "_pendingCallbacks": null,
    "_pendingElement": null,
    "_pendingForceUpdate": false,
    "_pendingReplaceState": false,
    "_pendingStateQueue": null,
    "_renderedComponent": NoopInternalComponent {
      "_currentElement": <div>
        <button
                className="btn btn-default"
                onClick={[Function]}>
                Create Account
        </button>
</div>,
      "_debugID": 8,
      "_renderedOutput": <div>
        <button
                className="btn btn-default"
                onClick={[Function]}>
                Create Account
        </button>
</div>,
    },
    "_renderedNodeType": 0,
    "_rootNodeID": 0,
    "_topLevelWrapper": null,
    "_updateBatchNumber": null,
    "_warnedAboutRefsInRender": false,
  },
  "context": Object {},
  "displayEmailErrors": [Function],
  "handleEmail": [Function],
  "handlePassword": [Function],
  "handleSubmit": [Function],
  "hideModal": [Function],
  "props": Object {},
  "refs": Object {},
  "showModal": [Function],
  "state": Object {
    "email": "",
    "emailValidations": Object {
      "errors": Array [],
      "valid": false,
    },
    "password": "Pa$$word!",
    "passwordValidations": Object {
      "errors": Array [],
      "valid": true,
    },
    "showCreateAccount": false,
    "showEmailErrors": false,
  },
  "updater": Object {
    "enqueueCallback": [Function],
    "enqueueCallbackInternal": [Function],
    "enqueueElementInternal": [Function],
    "enqueueForceUpdate": [Function],
    "enqueueReplaceState": [Function],
    "enqueueSetState": [Function],
    "isMounted": [Function],
    "validateCallback": [Function],
  },
}
`;

exports[`<App /> should handle the form submit 1`] = `
App {
  "_reactInternalInstance": ShallowComponentWrapper {
    "_calledComponentWillUnmount": false,
    "_compositeType": 0,
    "_context": Object {},
    "_currentElement": <App />,
    "_debugID": 11,
    "_hostContainerInfo": null,
    "_hostParent": null,
    "_instance": [Circular],
    "_mountOrder": 6,
    "_pendingCallbacks": null,
    "_pendingElement": null,
    "_pendingForceUpdate": false,
    "_pendingReplaceState": false,
    "_pendingStateQueue": null,
    "_renderedComponent": NoopInternalComponent {
      "_currentElement": <div>
        <p>
                Successfully created user
        </p>
</div>,
      "_debugID": 12,
      "_renderedOutput": <div>
        <p>
                Successfully created user
        </p>
</div>,
    },
    "_renderedNodeType": 0,
    "_rootNodeID": 0,
    "_topLevelWrapper": null,
    "_updateBatchNumber": null,
    "_warnedAboutRefsInRender": false,
  },
  "context": Object {},
  "displayEmailErrors": [Function],
  "handleEmail": [Function],
  "handlePassword": [Function],
  "handleSubmit": [Function],
  "hideModal": [Function],
  "props": Object {},
  "refs": Object {},
  "showModal": [Function],
  "state": Object {
    "email": "",
    "emailValidations": Object {
      "errors": Array [],
      "valid": false,
    },
    "password": "",
    "passwordValidations": Object {
      "errors": Array [
        "Password must contain one of the following symbols: !#$%&\'*+-/=?^_\`{|}~",
        "Password must be at least 8 characters",
      ],
      "valid": false,
    },
    "showCreateAccount": false,
    "showEmailErrors": false,
    "userCreated": true,
  },
  "updater": Object {
    "enqueueCallback": [Function],
    "enqueueCallbackInternal": [Function],
    "enqueueElementInternal": [Function],
    "enqueueForceUpdate": [Function],
    "enqueueReplaceState": [Function],
    "enqueueSetState": [Function],
    "isMounted": [Function],
    "validateCallback": [Function],
  },
}
`;

exports[`<App /> should know when to display email errors 1`] = `
App {
  "_reactInternalInstance": ShallowComponentWrapper {
    "_calledComponentWillUnmount": false,
    "_compositeType": 0,
    "_context": Object {},
    "_currentElement": <App />,
    "_debugID": 9,
    "_hostContainerInfo": null,
    "_hostParent": null,
    "_instance": [Circular],
    "_mountOrder": 5,
    "_pendingCallbacks": null,
    "_pendingElement": null,
    "_pendingForceUpdate": false,
    "_pendingReplaceState": false,
    "_pendingStateQueue": null,
    "_renderedComponent": NoopInternalComponent {
      "_currentElement": <div>
        <button
                className="btn btn-default"
                onClick={[Function]}>
                Create Account
        </button>
</div>,
      "_debugID": 10,
      "_renderedOutput": <div>
        <button
                className="btn btn-default"
                onClick={[Function]}>
                Create Account
        </button>
</div>,
    },
    "_renderedNodeType": 0,
    "_rootNodeID": 0,
    "_topLevelWrapper": null,
    "_updateBatchNumber": null,
    "_warnedAboutRefsInRender": false,
  },
  "context": Object {},
  "displayEmailErrors": [Function],
  "handleEmail": [Function],
  "handlePassword": [Function],
  "handleSubmit": [Function],
  "hideModal": [Function],
  "props": Object {},
  "refs": Object {},
  "showModal": [Function],
  "state": Object {
    "email": "",
    "emailValidations": Object {
      "errors": Array [],
      "valid": false,
    },
    "password": "",
    "passwordValidations": Object {
      "errors": Array [
        "Password must contain one of the following symbols: !#$%&\'*+-/=?^_\`{|}~",
        "Password must be at least 8 characters",
      ],
      "valid": false,
    },
    "showCreateAccount": false,
    "showEmailErrors": true,
  },
  "updater": Object {
    "enqueueCallback": [Function],
    "enqueueCallbackInternal": [Function],
    "enqueueElementInternal": [Function],
    "enqueueForceUpdate": [Function],
    "enqueueReplaceState": [Function],
    "enqueueSetState": [Function],
    "isMounted": [Function],
    "validateCallback": [Function],
  },
}
`;

exports[`<App /> should show and hide the modal 1`] = `
App {
  "_reactInternalInstance": ShallowComponentWrapper {
    "_calledComponentWillUnmount": false,
    "_compositeType": 0,
    "_context": Object {},
    "_currentElement": <App />,
    "_debugID": 3,
    "_hostContainerInfo": null,
    "_hostParent": null,
    "_instance": [Circular],
    "_mountOrder": 2,
    "_pendingCallbacks": null,
    "_pendingElement": null,
    "_pendingForceUpdate": false,
    "_pendingReplaceState": false,
    "_pendingStateQueue": null,
    "_renderedComponent": NoopInternalComponent {
      "_currentElement": <div>
        <button
                className="btn btn-default"
                onClick={[Function]}>
                Create Account
        </button>
        <Modal>
                <h2>
                        Create Account
                </h2>
                <Unknown
                        closeModal={[Function]}
                        displayEmailErrors={[Function]}
                        email=""
                        formErrors={
                                Array [
                                        "Password must contain one of the following symbols: !#$%&\'*+-/=?^_\`{|}~",
                                        "Password must be at least 8 characters",
                                      ]
                        }
                        handleEmail={[Function]}
                        handlePassword={[Function]}
                        handleSubmit={[Function]}
                        password="" />
                <button
                        className="btn btn-default close-modal"
                        onClick={[Function]}>
                        Close
                </button>
        </Modal>
</div>,
      "_debugID": 4,
      "_renderedOutput": <div>
        <button
                className="btn btn-default"
                onClick={[Function]}>
                Create Account
        </button>
        <Modal>
                <h2>
                        Create Account
                </h2>
                <Unknown
                        closeModal={[Function]}
                        displayEmailErrors={[Function]}
                        email=""
                        formErrors={
                                Array [
                                        "Password must contain one of the following symbols: !#$%&\'*+-/=?^_\`{|}~",
                                        "Password must be at least 8 characters",
                                      ]
                        }
                        handleEmail={[Function]}
                        handlePassword={[Function]}
                        handleSubmit={[Function]}
                        password="" />
                <button
                        className="btn btn-default close-modal"
                        onClick={[Function]}>
                        Close
                </button>
        </Modal>
</div>,
    },
    "_renderedNodeType": 0,
    "_rootNodeID": 0,
    "_topLevelWrapper": null,
    "_updateBatchNumber": null,
    "_warnedAboutRefsInRender": false,
  },
  "context": Object {},
  "displayEmailErrors": [Function],
  "handleEmail": [Function],
  "handlePassword": [Function],
  "handleSubmit": [Function],
  "hideModal": [Function],
  "props": Object {},
  "refs": Object {},
  "showModal": [Function],
  "state": Object {
    "email": "",
    "emailValidations": Object {
      "errors": Array [],
      "valid": false,
    },
    "password": "",
    "passwordValidations": Object {
      "errors": Array [
        "Password must contain one of the following symbols: !#$%&\'*+-/=?^_\`{|}~",
        "Password must be at least 8 characters",
      ],
      "valid": false,
    },
    "showCreateAccount": true,
    "showEmailErrors": false,
  },
  "updater": Object {
    "enqueueCallback": [Function],
    "enqueueCallbackInternal": [Function],
    "enqueueElementInternal": [Function],
    "enqueueForceUpdate": [Function],
    "enqueueReplaceState": [Function],
    "enqueueSetState": [Function],
    "isMounted": [Function],
    "validateCallback": [Function],
  },
}
`;

exports[`<App /> should show and hide the modal 2`] = `
App {
  "_reactInternalInstance": ShallowComponentWrapper {
    "_calledComponentWillUnmount": false,
    "_compositeType": 0,
    "_context": Object {},
    "_currentElement": <App />,
    "_debugID": 3,
    "_hostContainerInfo": null,
    "_hostParent": null,
    "_instance": [Circular],
    "_mountOrder": 2,
    "_pendingCallbacks": null,
    "_pendingElement": null,
    "_pendingForceUpdate": false,
    "_pendingReplaceState": false,
    "_pendingStateQueue": null,
    "_renderedComponent": NoopInternalComponent {
      "_currentElement": <div>
        <button
                className="btn btn-default"
                onClick={[Function]}>
                Create Account
        </button>
</div>,
      "_debugID": 4,
      "_renderedOutput": <div>
        <button
                className="btn btn-default"
                onClick={[Function]}>
                Create Account
        </button>
</div>,
    },
    "_renderedNodeType": 0,
    "_rootNodeID": 0,
    "_topLevelWrapper": null,
    "_updateBatchNumber": null,
    "_warnedAboutRefsInRender": false,
  },
  "context": Object {},
  "displayEmailErrors": [Function],
  "handleEmail": [Function],
  "handlePassword": [Function],
  "handleSubmit": [Function],
  "hideModal": [Function],
  "props": Object {},
  "refs": Object {},
  "showModal": [Function],
  "state": Object {
    "email": "",
    "emailValidations": Object {
      "errors": Array [],
      "valid": false,
    },
    "password": "",
    "passwordValidations": Object {
      "errors": Array [
        "Password must contain one of the following symbols: !#$%&\'*+-/=?^_\`{|}~",
        "Password must be at least 8 characters",
      ],
      "valid": false,
    },
    "showCreateAccount": false,
    "showEmailErrors": false,
  },
  "updater": Object {
    "enqueueCallback": [Function],
    "enqueueCallbackInternal": [Function],
    "enqueueElementInternal": [Function],
    "enqueueForceUpdate": [Function],
    "enqueueReplaceState": [Function],
    "enqueueSetState": [Function],
    "isMounted": [Function],
    "validateCallback": [Function],
  },
}
`;
